build/item.rs

```diff
@@ -0,0 +1,170 @@
+        .iter()
+        .map(|i| {
+            let id = &i.id;
+            let name_ident = ident(&i.name.to_pascal_case());
+
+            quote! {
+                Item::#name_ident => #id,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let item_from_raw_id_arms = items
+        .iter()
+        .map(|i| {
+            let id = &i.id;
+            let name_ident = ident(&i.name.to_pascal_case());
+
+            quote! {
+                #id => Some(Item::#name_ident),
+            }
+        })
+        .collect::<TokenStream>();
+
+    let item_to_max_count_arms = items
+        .iter()
+        .map(|i| {
+            let name_ident = ident(&i.name.to_pascal_case());
+            let max_count = i.max_stack;
+
+            quote! {
+                Item::#name_ident => #max_count,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let items_varients = items
+        .iter()
+        .map(|i| ident(i.name.to_pascal_case()))
+        .collect::<Vec<_>>();
+
+    Ok(quote! {
+        /// Represents an item from game
+        #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+        pub enum Item {
+            #(#item_varients,)*
+        }
+
+        impl Item {
+            /// Constructs a item from a raw item ID.
+            ///
+            /// If the given ID is invalid, `None` is returned.
+            pub const fn from_raw(id: u16) -> Option<Self> {
+                match id {
+                    #item_from_raw_id_arms
+                    _ => None
+                }
+
+                //if id <= #items_count.try_into().unwrap() {
+                //    Some(Self(id))
+                //} else {
+                //    None
+                //}
+            }
+
+            /// Gets the raw item ID from the item
+            pub const fn to_raw(self) -> u16 {
+                match self {
+                    #item_to_raw_id_arms
+                }
+            }
+
+            /// Construct an item for its snake_case name.
+            ///
+            /// Returns `None` if the name is invalid.
+            // Added to make it the same as BlockKind
+            #[allow(clippy::should_implement_trait)]
+            pub fn from_str(name: &str) -> Option<Item> {
+                match name {
+                    #item_from_str_arms
+                    _ => None
+                }
+            }
+
+            /// Get the snake_case name of this block kind.
+            pub const fn to_str(self) -> &'static str {
+                match self {
+                    #item_to_str_arms
+                }
+            }
+
+            /// Gets the translation key of an item.
+            pub const fn translation_key(self) -> &'static str {
+                match self {
+                    #item_to_translation_key_arms
+                }
+            }
+
+            /// Returns the max stack count
+            pub const fn get_max_count(self) -> i16 {
```

Comment:
nit: should this be `usize`?

Reply:
I thought about just using an `u8` since the count shouldn't get above 64, but in the Minecraft Code it's stores as an int, so i thought it would be best to use an `i16` so they were compatible.

Reply:
The Minecraft protocol uses a byte to represent item counts so making `max_stack` a `u8` should be fine. There will be a serde error at build time if the value is out of range for some reason.

Also, since this function calls it `max_count` we should rename `max_stack` to `max_count`.